tsdb/head.go

```diff
@@ -80,27 +83,44 @@ type Head struct {
 	cardinalityMutex      sync.Mutex
 	cardinalityCache      *index.PostingsStats // Posting stats cache which will expire after 30sec.
 	lastPostingsStatsCall time.Duration        // Last posting stats call (PostingsCardinalityStats()) time for caching.
+
+	// chunkDiskMapper is used to write and ready Head chunks to/from disk.
+	chunkDiskMapper *chunks.ChunkDiskMapper
+	// chunkDirRoot is the parent directory of the chunks directory.
+	chunkDirRoot string
+
+	// seriesReplayCorruption is set to true during WAL replay if there was corruption while
+	// loading the series from WAL.
+	seriesReplayCorruption bool
+	// refSeries a map of series ref to the actual series loaded during WAL replay.
+	refSeries map[uint64]*memSeries
+	// multiRef maps the duplicate series ref to the one in the memory.
+	// This can happen when there are multiple series records for the same series
+	// in the WAL. This mapping is required to map the samples with wrong ref to the actual series ref in memory.
+	multiRef map[uint64]uint64
 }
 
 type headMetrics struct {
-	activeAppenders         prometheus.Gauge
-	series                  prometheus.GaugeFunc
-	seriesCreated           prometheus.Counter
-	seriesRemoved           prometheus.Counter
-	seriesNotFound          prometheus.Counter
-	chunks                  prometheus.Gauge
-	chunksCreated           prometheus.Counter
-	chunksRemoved           prometheus.Counter
-	gcDuration              prometheus.Summary
-	samplesAppended         prometheus.Counter
-	walTruncateDuration     prometheus.Summary
-	walCorruptionsTotal     prometheus.Counter
-	headTruncateFail        prometheus.Counter
-	headTruncateTotal       prometheus.Counter
-	checkpointDeleteFail    prometheus.Counter
-	checkpointDeleteTotal   prometheus.Counter
-	checkpointCreationFail  prometheus.Counter
-	checkpointCreationTotal prometheus.Counter
+	activeAppenders          prometheus.Gauge
+	series                   prometheus.GaugeFunc
+	seriesCreated            prometheus.Counter
+	seriesRemoved            prometheus.Counter
+	seriesNotFound           prometheus.Counter
+	chunks                   prometheus.Gauge
+	chunksCreated            prometheus.Counter
+	chunksRemoved            prometheus.Counter
+	gcDuration               prometheus.Summary
+	samplesAppended          prometheus.Counter
+	outOfOrderSamples        prometheus.Counter
```

Comment:
This is not registered.

Reply:
hm, wait isn't there a test for this metric somewhere in the tests? how did we miss that?

Reply:
There is no test for this. I can write one if it is necessary.

Reply:
thanos has tests to see if all metrics are registered

Reply:
I will add a test. I had missed incrementing the metric when we return `storage.ErrOutOfBounds`

Reply:
I added a test now, and I have split the metric into 2, 1 for out of order (within compaction window but out of order), 1 for out of bound (beyond the compaction window)

Reply:
I had this in mind but thought I will follow up with another PR after this to keep the diff size smaller. I would still prefer to do it in a separate PR.

Reply:
Cool, up to you